ddtrace/contrib/kafka/patch.py

```diff
@@ -0,0 +1,124 @@
+        if not pin or not pin.enabled():
+            return func(*args, **kwargs)
+
+        with pin.tracer.trace(
+            kafkax.PRODUCE,
+            service=trace_utils.ext_service(pin, config.kafka),
+            span_type=SpanTypes.WORKER,
+        ) as span:
+            span.set_tag_str("messaging.system", kafkax.SERVICE)
+            span.set_tag_str(COMPONENT, config.kafka.integration_name)
+            span.set_tag_str(SPAN_KIND, SpanKind.PRODUCER)
+            span.set_tag_str(kafkax.TOPIC, topic)
+            span.set_tag_str(kafkax.MESSAGE_KEY, ensure_text(message_key))
+            span.set_tag(kafkax.PARTITION, partition)
+            span.set_tag_str(kafkax.TOMBSTONE, str(value is None))
+            span.set_tag(SPAN_MEASURED_KEY)
+            rate = config.kafka.get_analytics_sample_rate()
+            if rate is not None:
+                span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, rate)
+            return func(*args, **kwargs)
+
+    # in older versions of confluent_kafka, bool(Producer()) evaluates to False,
+    # which makes the Pin functionality ignore it.
+    def __bool__(self):
+        return True
+
+    __nonzero__ = __bool__
+
+
+class TracedConsumer(ObjectProxy):
+    def __init__(self, *args, **kwargs):
+        consumer = _Consumer(*args, **kwargs)
+        super(TracedConsumer, self).__init__(consumer)
+        Pin().onto(self)
+
+    def poll(self, *args, **kwargs):
+        func = self.__wrapped__.poll
+        pin = Pin.get_from(self)
+        if not pin or not pin.enabled():
+            return func(*args, **kwargs)
+
+        with pin.tracer.trace(
+            kafkax.CONSUME,
+            service=trace_utils.ext_service(pin, config.kafka),
+            span_type=SpanTypes.WORKER,
+        ) as span:
+            message = func(*args, **kwargs)
+            span.set_tag_str("messaging.system", kafkax.SERVICE)
+            span.set_tag_str(COMPONENT, config.kafka.integration_name)
+            span.set_tag_str(SPAN_KIND, SpanKind.CONSUMER)
+            span.set_tag_str(kafkax.RECEIVED_MESSAGE, str(message is not None))
+            if message is not None:
+                message_key = message.key() or ""
+                span.set_tag_str(kafkax.TOPIC, message.topic())
+                span.set_tag_str(kafkax.MESSAGE_KEY, ensure_text(message_key))
+                span.set_tag(kafkax.PARTITION, message.partition())
+                span.set_tag_str(kafkax.TOMBSTONE, str(len(message) == 0))
+            span.set_tag(SPAN_MEASURED_KEY)
+            rate = config.kafka.get_analytics_sample_rate()
+            if rate is not None:
+                span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, rate)
+            return message
+
+
+def patch():
+    if getattr(confluent_kafka, "_datadog_patch", False):
+        return
+    setattr(confluent_kafka, "_datadog_patch", True)
+
+    setattr(confluent_kafka, "Producer", TracedProducer)
+    setattr(confluent_kafka, "Consumer", TracedConsumer)
```

Comment:
Instead of patching the Producer and Consumer classes can we patch the methods directly. Patching classes leads to some edge cases. It has bit us before with the TracedMongoClient.

If a user imports `confluent_kafka.Producer`  and/or `confluent_kafka.Consumer` then calls ddtrace.patch(kafka=True). We won't patch any Producer or Consumer classes that have already been imported.

Producer is not patched:
```
>>> from confluent_kafka import Producer, Consumer
>>> from ddtrace import patch
>>> patch(kafka=True)
>>> pd = Producer(dict())
%5|1679066975.054|CONFWARN|rdkafka#producer-1| [thrd:app]: No `bootstrap.servers` configured: client will not be able to connect to Kafka cluster
>>> type(pd)
<class 'cimpl.Producer'>
```

Producer is patched:
```
 >>> from ddtrace import patch
>>> patch(kafka=True)
>>> from confluent_kafka import Producer, Consumer
>>> pd = Producer(dict())
%5|1679066848.217|CONFWARN|rdkafka#producer-1| [thrd:app]: No `bootstrap.servers` configured: client will not be able to connect to Kafka cluster
>>> type(pd)
<class 'ddtrace.contrib.kafka.patch.TracedProducer'>
```

Reply:
@mabdinur unfortunately these methods can't be patched directly because they are [implemented in C](https://github.com/confluentinc/confluent-kafka-python/blob/bbe1225dc7d36df2c7e3e0f7476bb283fabcd1ff/src/confluent_kafka/src/confluent_kafka.c#L2698) and [directly imported](https://github.com/confluentinc/confluent-kafka-python/blob/bbe1225dc7d36df2c7e3e0f7476bb283fabcd1ff/src/confluent_kafka/__init__.py#L24).

This pull request implements a similar solution to [OpenTelemetry's](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/2629c121610693ad3add1e84e94d237733271403/instrumentation/opentelemetry-instrumentation-confluent-kafka/src/opentelemetry/instrumentation/confluent_kafka/__init__.py#L258).

Reply:
Isn't our patch guidance to patch before importing? AFAIK our `patch()` or `patch_all()` works by registering the patch hook to the module import, meaning users have to patch before importing `confluent_kafka`.